File: java/src/main/java/com/securityeventlogger/SecurityEventLogger.java
package com.securityeventlogger;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.JsonNode;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;

/**
 * OWASP Security Event Logger for Java
 * A library for logging security events according to the OWASP Logging Vocabulary
 */
public class SecurityEventLogger {
    private final String appId;
    private final Map<String, EventDefinition> eventsDef;
    private final ObjectMapper objectMapper;

    /**
     * Initialize the security event logger
     * @param appId Application identifier
     * @throws IOException if the events file cannot be read
     */
    public SecurityEventLogger(String appId) throws IOException {
        this(appId, null);
    }

    /**
     * Initialize the security event logger
     * @param appId Application identifier
     * @param eventsFile Path to security_events.json file (optional)
     * @throws IOException if the events file cannot be read
     */
    public SecurityEventLogger(String appId, String eventsFile) throws IOException {
        this.appId = appId;
        this.objectMapper = new ObjectMapper();
        this.eventsDef = new HashMap<>();

        // Load events definition
        JsonNode rootNode;
        if (eventsFile == null || eventsFile.isEmpty()) {
            // Load from classpath or default location
            InputStream is = getClass().getClassLoader().getResourceAsStream("security_events.json");
            if (is == null) {
                // Try parent directory
                eventsFile = "../security_events.json";
                rootNode = objectMapper.readTree(Files.newInputStream(Paths.get(eventsFile)));
            } else {
                rootNode = objectMapper.readTree(is);
            }
        } else {
            rootNode = objectMapper.readTree(Files.newInputStream(Paths.get(eventsFile)));
        }

        JsonNode eventsNode = rootNode.get("events");
        Iterator<Map.Entry<String, JsonNode>> fields = eventsNode.fields();
        while (fields.hasNext()) {
            Map.Entry<String, JsonNode> entry = fields.next();
            String eventType = entry.getKey();
            JsonNode eventNode = entry.getValue();

            EventDefinition eventDef = new EventDefinition(
                eventNode.get("category").asText(),
                eventNode.get("prefix").asText(),
                eventNode.get("level").asText(),
                eventNode.get("description").asText()
            );

            JsonNode paramsNode = eventNode.get("parameters");
            if (paramsNode != null && paramsNode.isArray()) {
                for (JsonNode param : paramsNode) {
                    eventDef.addParameter(param.asText());
                }
            }

            eventsDef.put(eventType, eventDef);
        }
    }

    /**
     * Log a security event
     * @param eventType Type of event (e.g., 'authn_login_success')
     * @param params List of parameters for the event
     * @param options Optional fields
     * @return The log event as a Map
     * @throws IllegalArgumentException if the event type is unknown
     */
    public Map<String, Object> logEvent(String eventType, List<String> params, Map<String, String> options) {
        if (!eventsDef.containsKey(eventType)) {
            throw new IllegalArgumentException("Unknown event type: " + eventType);
        }

        EventDefinition eventDef = eventsDef.get(eventType);

        // Build event string
        StringBuilder eventStr = new StringBuilder(eventType);
        if (params != null && !params.isEmpty()) {
            eventStr.append(":").append(String.join(",", params));
        }

        // Build log entry
        Map<String, Object> logEntry = new LinkedHashMap<>();
        logEntry.put("datetime", ZonedDateTime.now().format(DateTimeFormatter.ISO_OFFSET_DATE_TIME));
        logEntry.put("appid", appId);
        logEntry.put("event", eventStr.toString());
        logEntry.put("level", eventDef.getLevel());

        String description = (options != null && options.containsKey("description")) 
            ? options.get("description") 
            : eventDef.getDescription();
        logEntry.put("description", description);

        // Add optional fields
        if (options != null) {
            String[] optionalFields = {
                "useragent", "source_ip", "host_ip", "hostname",
                "protocol", "port", "request_uri", "request_method",
                "region", "geo"
            };

            Set<String> knownFields = new HashSet<>(Arrays.asList(optionalFields));
            knownFields.add("description");

            for (Map.Entry<String, String> entry : options.entrySet()) {
                String key = entry.getKey();
                String value = entry.getValue();

                if (!key.equals("description") && value != null && !value.isEmpty()) {
                    logEntry.put(key, value);
                }
            }
        }

        return logEntry;
    }

    /**
     * Get information about an event type
     * @param eventType Type of event
     * @return Event information
     * @throws IllegalArgumentException if the event type is unknown
     */
    public EventDefinition getEventInfo(String eventType) {
        if (!eventsDef.containsKey(eventType)) {
            throw new IllegalArgumentException("Unknown event type: " + eventType);
        }
        return eventsDef.get(eventType);
    }

    /**
     * List available event types
     * @param category Optional category filter
     * @return List of event type names
     */
    public List<String> listEvents(String category) {
        List<String> events = new ArrayList<>();
        for (Map.Entry<String, EventDefinition> entry : eventsDef.entrySet()) {
            if (category == null || category.isEmpty() || entry.getValue().getCategory().equals(category)) {
                events.add(entry.getKey());
            }
        }
        return events;
    }

    /**
     * EventDefinition represents a security event definition
     */
    public static class EventDefinition {
        private final String category;
        private final String prefix;
        private final String level;
        private final String description;
        private final List<String> parameters;

        public EventDefinition(String category, String prefix, String level, String description) {
            this.category = category;
            this.prefix = prefix;
            this.level = level;
            this.description = description;
            this.parameters = new ArrayList<>();
        }

        public void addParameter(String parameter) {
            this.parameters.add(parameter);
        }

        public String getCategory() { return category; }
        public String getPrefix() { return prefix; }
        public String getLevel() { return level; }
        public String getDescription() { return description; }
        public List<String> getParameters() { return parameters; }
    }
}

File: golang/security_event_logger.go
package securityeventlogger

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"path/filepath"
	"runtime"
	"strings"
	"time"

	"gopkg.in/yaml.v3"
)

// EventDefinition represents a security event definition
type EventDefinition struct {
	Category    string   `yaml:"category"`
	Prefix      string   `yaml:"prefix"`
	Level       string   `yaml:"level"`
	Description string   `yaml:"description"`
	Parameters  []string `yaml:"parameters"`
}

// EventsData represents the structure of the events YAML file
type EventsData struct {
	Version string                     `yaml:"version"`
	Events  map[string]EventDefinition `yaml:"events"`
}

// LogEntry represents a security log entry
type LogEntry struct {
	DateTime      string            `json:"datetime"`
	AppID         string            `json:"appid"`
	Event         string            `json:"event"`
	Level         string            `json:"level"`
	Description   string            `json:"description"`
	UserAgent     string            `json:"useragent,omitempty"`
	SourceIP      string            `json:"source_ip,omitempty"`
	HostIP        string            `json:"host_ip,omitempty"`
	Hostname      string            `json:"hostname,omitempty"`
	Protocol      string            `json:"protocol,omitempty"`
	Port          string            `json:"port,omitempty"`
	RequestURI    string            `json:"request_uri,omitempty"`
	RequestMethod string            `json:"request_method,omitempty"`
	Region        string            `json:"region,omitempty"`
	Geo           string            `json:"geo,omitempty"`
	Custom        map[string]string `json:"-"`
}

// SecurityEventLogger is the main logger struct
type SecurityEventLogger struct {
	appID     string
	eventsDef map[string]EventDefinition
}

// NewSecurityEventLogger creates a new security event logger
func NewSecurityEventLogger(appID string, eventsFile string) (*SecurityEventLogger, error) {
	if eventsFile == "" {
		// Get the path relative to this file
		_, filename, _, _ := runtime.Caller(0)
		eventsFile = filepath.Join(filepath.Dir(filepath.Dir(filename)), "security_events.yaml")
	}

	data, err := ioutil.ReadFile(eventsFile)
	if err != nil {
		return nil, fmt.Errorf("failed to read events file: %w", err)
	}

	var eventsData EventsData
	if err := yaml.Unmarshal(data, &eventsData); err != nil {
		return nil, fmt.Errorf("failed to parse events file: %w", err)
	}

	return &SecurityEventLogger{
		appID:     appID,
		eventsDef: eventsData.Events,
	}, nil
}

// LogEvent logs a security event
func (l *SecurityEventLogger) LogEvent(eventType string, params []string, options map[string]string) (*LogEntry, error) {
	eventDef, exists := l.eventsDef[eventType]
	if !exists {
		return nil, fmt.Errorf("unknown event type: %s", eventType)
	}

	// Build event string
	eventStr := eventType
	if len(params) > 0 {
		eventStr += ":" + strings.Join(params, ",")
	}

	// Get description
	description := eventDef.Description
	if desc, ok := options["description"]; ok {
		description = desc
	}

	// Build log entry
	entry := &LogEntry{
		DateTime:    time.Now().UTC().Format(time.RFC3339),
		AppID:       l.appID,
		Event:       eventStr,
		Level:       eventDef.Level,
		Description: description,
		Custom:      make(map[string]string),
	}

	// Add optional fields
	if val, ok := options["useragent"]; ok {
		entry.UserAgent = val
	}
	if val, ok := options["source_ip"]; ok {
		entry.SourceIP = val
	}
	if val, ok := options["host_ip"]; ok {
		entry.HostIP = val
	}
	if val, ok := options["hostname"]; ok {
		entry.Hostname = val
	}
	if val, ok := options["protocol"]; ok {
		entry.Protocol = val
	}
	if val, ok := options["port"]; ok {
		entry.Port = val
	}
	if val, ok := options["request_uri"]; ok {
		entry.RequestURI = val
	}
	if val, ok := options["request_method"]; ok {
		entry.RequestMethod = val
	}
	if val, ok := options["region"]; ok {
		entry.Region = val
	}
	if val, ok := options["geo"]; ok {
		entry.Geo = val
	}

	// Store custom fields
	knownFields := map[string]bool{
		"description": true, "useragent": true, "source_ip": true,
		"host_ip": true, "hostname": true, "protocol": true,
		"port": true, "request_uri": true, "request_method": true,
		"region": true, "geo": true,
	}
	for key, val := range options {
		if !knownFields[key] {
			entry.Custom[key] = val
		}
	}

	return entry, nil
}

// GetEventInfo returns information about an event type
func (l *SecurityEventLogger) GetEventInfo(eventType string) (*EventDefinition, error) {
	eventDef, exists := l.eventsDef[eventType]
	if !exists {
		return nil, fmt.Errorf("unknown event type: %s", eventType)
	}
	return &eventDef, nil
}

// ListEvents returns a list of available event types
func (l *SecurityEventLogger) ListEvents(category string) []string {
	var events []string
	for eventType, eventDef := range l.eventsDef {
		if category == "" || eventDef.Category == category {
			events = append(events, eventType)
		}
	}
	return events
}

// MarshalJSON implements custom JSON marshaling for LogEntry
func (e *LogEntry) MarshalJSON() ([]byte, error) {
	type Alias LogEntry
	aux := &struct {
		*Alias
	}{
		Alias: (*Alias)(e),
	}

	// Convert to map to include custom fields
	m := make(map[string]interface{})
	data, _ := json.Marshal(aux)
	json.Unmarshal(data, &m)

	// Add custom fields
	for k, v := range e.Custom {
		m[k] = v
	}

	return json.Marshal(m)
}
